internal/validators/namespace.go

```diff
@@ -107,6 +118,46 @@ func (v *Namespace) handle(req *nsRequest) admission.Response {
 	return allow("")
 }
 
+// illegalTreeLabel checks if tree labels are being created or modified
+// by any user or service account since only HNC service account is
+// allowed to do so
+func (v *Namespace) illegalTreeLabel(req *nsRequest) admission.Response {
+	msg := "Only HNC Service account can modify tree labels"
+
+	if req.op == k8sadm.Delete {
+		return allow("")
+	}
+
+	oldLabels := map[string]string{}
+	if req.oldns != nil {
+		oldLabels = req.oldns.Labels
+	}
+
+	for key, val := range req.ns.Labels {
+		if oldVal, ok := oldLabels[key]; !ok {
```

Comment:
I might restructure this loop a bit. E.g.:

```
if !strings.Contains(...) {
  continue
}
if oldVal, exists := oldLabels[key]; !exists || oldVal != val {
  return deny(...)
}
```

This has less duplication and IMO is a bit easier to read.